Dev options and Allow header

dai65527 opened the pull request:
OPTIONSに対するレスポンスとAllowヘッダ実装しましたので、ご確認お願いします。

### 内容
- configのlimit_exceptを確認して、Allowヘッダを作成します。
- OPTIONSリクエストに対しては、Allowヘッダを追加して、ボディのないレスポンスを返すようにしています。
- 405 Method Not Allowed を返す際に、OPTIONSが有効な場合のみAllowヘッダを追加するようにしました。
  - OPTIONSが有効な場合のみ、というのはセキュリティ上の配慮です。（NGINXも基本Allowヘッダは返さない）

参考：https://developer.mozilla.org/ja/docs/Web/HTTP/Headers/Allow

### 動作確認

config (fileupload.conf)

```
server {
    root ./html;
    listen 8888;

    limit_except GET POST OPTIONS;
    upload_pass /upload;
    upload_store /uploadstore/;
    
    location /location/ {
        limit_except GET OPTIONS;
        upload_pass /upload;
        upload_store /uploadstore/;
    }
}
```

OPTIONS

```
$ curl -i -X OPTIONS localhost:8888
HTTP/1.1 200 OK
Server: nginDX
Date: Wed, 21 Apr 2021 00:53:37 GMT
Allow: GET, HEAD, POST, OPTIONS
Content-Length: 0
Connection: keep-alive

```

405 Method not allowed

```
$ curl -i -X HOGE localhost:8888 
HTTP/1.1 405 Method Not Allowed
Server: nginDX
Date: Wed, 21 Apr 2021 00:53:04 GMT
Content-Type: text/html
Allow: GET, HEAD, POST, OPTIONS
Content-Length: 184
Connection: keep-alive

<html>
<head><title>405 Method Not Allowed</title></ head>
<body bgcolor = "white">
<center><h1>405 Method Not Allowed</h1></center>
<hr><center> nginDX </center>
</body></html>
```

dai65527 committed:
Add: Response to options

dai65527 committed:
Update Session::createErrorResponse

- to add "Allow" header when 405 method not allowed and OPTIONS request allowed

dai65527 committed:
mod fileupload conf to test OPTIONS

TrainCoffeeBoy commented on Session.cpp:
```diff
@@ -1258,6 +1269,57 @@ int Session::writeToFile() {
+  if (limit_except & HTTP_PUT) {
+    allowed_methods.append(allowed_methods.empty() ? "PUT" : ", PUT");
+  }
+  if (limit_except & HTTP_DELETE) {
+    allowed_methods.append(allowed_methods.empty() ? "DELETE" : ", DELETE");
+  }
+  if (limit_except & HTTP_DELETE) {
+    allowed_methods.append(allowed_methods.empty() ? "DELETE" : ", HEAD");
```
ここはL1309がHEADの方ですかね？configにDELETEしかなくてもHEADも表示されるようになっているようです。確認をお願いします（セルフマージでOKです）。

dai65527 replied:
あっ…ありがとうございます…！！

TrainCoffeeBoy reviewed (changes requested):
createAllowHeaderのHEADのところだけ、確認お願いします。

dai65527 committed:
mod createAllowHeader

- rm typo
- add connect (CONNECT not to be implemented but for future development)

dai65527 committed:
Merge branch 'main' of https://github.com/dai65527/webserv into dev_options

dai65527 commented:
self merge!!
